pyroomacoustics/room.py

```diff
@@ -2189,15 +2212,230 @@ def extrude(self, height):
-        :param room: the room object
-        :return: the volume in cubic unit
-        """
+        # Throw an error if we are missing some hardware in the room
+        if len(self.sources) == 0:
+            raise ValueError("There are no sound sources in the room.")
+        if self.mic_array is None:
+            raise ValueError("There is no microphone in the room.")
 
-        return np.prod(self.shoebox_dim)
+        # number of mics and sources
+        M = self.mic_array.M
+        S = len(self.sources)
+
+        # compute all delays from M microphones to S sources, and 
+        # maximum signal length
+
+        max_sig_len = 0 
+
+        # TODO(FD) make this faster using broadcasting
+        delays = np.zeros((M, S), dtype=np.int)
+        for m in range(M):
+            for s in range(S):
+
+                source_delay = self.sources[s].delay * self.fs
+                mic_delay = self.get_delay(s, m)  * self.fs
+
+                #TODO(FD) use floor or round here? 
+                delay = int(source_delay + mic_delay)
+                delays[m, s] = delay
+
+                max_sig_len = max(delay + len(self.sources[s].signal), max_sig_len)
+
+        # TODO(FD) not sure if removing the zero-signal stretch poses problems
+        # for applications other than DOA. 
+        min_delay = delays.min()
+        max_sig_len -= min_delay
+        delays -= min_delay
+
+        # the array that will receive all the signals
+        premix_signals = np.zeros((S, M, max_sig_len))
+
+        # compute the signal at every microphone in the array
+        for m in np.arange(M):
+            for s in np.arange(S):
+                sig = self.sources[s].signal
+                if sig is None:
+                    continue
+
+                d = delays[m, s]
+                premix_signals[s, m, d:d+len(sig)] += sig
+
+        if callback_mix is not None:
+            # Execute user provided callback
+            signals = callback_mix(premix_signals, **callback_mix_kwargs)
+            self.sigma2_awgn = None
+
+        elif snr is not None:
+            # Normalize all signals so that
+            denom = np.std(premix_signals[:, reference_mic, :], axis=1)
+            premix_signals /= denom[:, None, None]
+            signals = np.sum(premix_signals, axis=0)
+
+            # Compute the variance of the microphone noise
+            self.sigma2_awgn = 10 ** (-snr / 10) * S
+
+        else:
+            signals = np.sum(premix_signals, axis=0)
+
+        # add white gaussian noise if necessary
+        if self.sigma2_awgn is not None:
+            signals += np.random.normal(0.0, np.sqrt(self.sigma2_awgn), signals.shape)
+
+        # record the signals in the microphones
+        self.mic_array.record(signals, self.fs)
+
+        if return_premix:
+            return premix_signals
+
+    # TODO(FD): not sure if we want the far-field option below? 
+    def get_delay(self, source_idx, mic_idx, mic_ref_idx=None):
```

Comment:
I have included the option to specify a reference mic, in which case the source is considered far-field and the delays are computed with respect to that reference. I am not sure if this is a desired feature, as it does not properly reflect the real physics of the setup.